ncm-download/src/main/perl/download.pm

```diff
@@ -306,12 +308,25 @@ sub retrieve
                 if (!POSIX::WIFEXITED($?) || POSIX::WEXITSTATUS($?) != 0) {
                     # Only warn as there may be attempts from multiple proxies.
                     $self->warn("curl failed (". ($?>>8) ."): $errs\nCommand = $proc");
+                    if (-e $tmpfn) {
+                        unless (unlink($tmpfn)) {
+                            $self->error("Unable to delete temporary file $tmpfn: $!");
+                        }
+                    }
                     return 0; # Curl barfed, so we return.
                 } else {
                     if ($CAF::Object::NoAction) {
                         $self->debug(1, "We have sucessfully pretended to retrieve a new copy of the file $fn");
                     } else {
-                        $self->debug(1, "We seem to have sucessfully retrieved a new copy of the file $fn");
+                        $self->debug(1, "We seem to have sucessfully retrieved a new copy of the file $fn to $tmpfn");
+                        if (rename($tmpfn, $fn)) {
+                            $self->debug(1, "Succesfully renamed temporary file $tmpfn to $fn");
+                        } else {
+                            $self->error("Unable to rename temporary file $tmpfn to $fn: $!");
+                            unless (unlink($tmpfn)) {
```

Comment:
use the same `if (-e ...)` logic as above; and even better, make a small private method for it